Fix: Discord user name and remove discriminator

SlySven opened the pull request:
#### Brief overview of PR changes/additions
Revises the UI so that ~~it is possible to specify a Discord "Discriminator" of just `"0"` which is the form that it will take once a user has migrated to the new user name format~~ *the Discord discriminator is completely removed, as Discord no longer uses it*.

Also warns if a user name that contains invalid character for the "new" format or any other discriminator is detect in the saved settings when they are loaded so that the end-user is advised to update them when they are forced to change their Discord username to the "new" style.

#### Motivation for adding to Mudlet
Discord's user name changes - as described here:
https://support.discord.com/hc/en-us/articles/12620128861463-New-Usernames-Display-Names
and here:
https://discord.com/blog/usernames
mean that users will have to change their Discord usernames and that will mean that the user identity reported back from the local Discord application via the Discord-RPC library to Mudlet will change and stop Mudlet from allowing a profile to use the Discord-RPC if it has set a restriction to the user name or discriminator. To aide in the process of switching over this PR will cause Mudlet to issue `[ WARN ]  - ` type messages on the main console for any profile where it detects that a user name has been set to limit the Discord-RPC library from being used with the profile AND that user name is of the form that is it an "old" rather than a "new" format or that a discriminator has been used. ~~The only valid discriminator in the new form is a single zero digit and having gone through the user name change for my main Discord account I can confer that that is what the Discord-RPC reports back - however the Mudlet code prior to this PR would not accept anything other than a full 4 digit number (or nothing at all for this value)~~.

#### Other info (issues closed, discussion etc)
The changes included in this PR are smaller and simpler than the ones I proposed in the (currently still a draft) PR #6835.

SlySven committed:
Fix: allow single zero digit discord discriminator

Also warns if a user name that contains invalid character for the "new"
format or any other discriminator is detect in the saved settings when they
are loaded so that the end-user is advised to update them when they are
forced to change their Discord username to the "new" style.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Fix_allow_singleZeroDigitDiscordDiscriminator

SlySven commented:
Marking this as a regression because Discord has changed things causing this aspect of the Discord-RPC interface that Mudlet provides to break (for me who uses it) - but only marking it as **medium** because of the size of the group of people affected!

SlySven committed:
Merge branch 'development' into Fix_allow_singleZeroDigitDiscordDiscriminator

SlySven committed:
Merge branch 'development' into Fix_allow_singleZeroDigitDiscordDiscriminator

SlySven commented:
CodeFactor is unhappy because I happened to modify something in a "complex method" - but I have no intention of making a massive unrelated change to refactor that method just to shut it up...

SlySven committed:
Merge branch 'development' into Fix_allow_singleZeroDigitDiscordDiscriminator

SlySven committed:
Merge branch 'development' into Fix_allow_singleZeroDigitDiscordDiscriminator

SlySven committed:
Merge branch 'development' into Fix_allow_singleZeroDigitDiscordDiscriminator

SlySven committed:
Merge branch 'development' into Fix_allow_singleZeroDigitDiscordDiscriminator

vadi2 commented:
Just a question... Discord made their changes back in May and we have not heard any complaints from Mudlet users at all now in December. Do we think anyone is using this functionality?

Kebap commented:
We do use (some of) the Discord integration in MG package, but I fail to understand from here or there #6835 what bug this PR is supposed to fix. Could you provide a failing example?

SlySven commented:
> Could you provide a failing example?

Have a non-empty "discriminator" here (this is the current `development` branch code, SHA-1 4b7fbb1f7ac62cf2e0df6af9822457e5ea2479c7 ):
![image](https://github.com/Mudlet/Mudlet/assets/6163092/36445f98-0738-4f74-a871-9852a1e5c425)

However this Discord account has been migrated to the new format - so this - my original settings for this account - doesn't work when they are as entered (the mention of Mudlet is a separate issue and not relevant for this PR):
![image](https://github.com/Mudlet/Mudlet/assets/6163092/af05223d-ad2d-4ea9-b356-146e1c33044a)

According to https://support-dev.discord.com/hc/en-us/articles/13667755828631 during the transition usernames will have a `#0` Discriminator - however it is not currently possible to set that in the current `development` code as there is a validator that only allows an empty string or exactly four digits (as the Discriminator used to be). Whilst you can enter a single zero:
![image](https://github.com/Mudlet/Mudlet/assets/6163092/55f68686-bd7c-4e1f-a496-55c9d36083eb)
it is rejected and not saved, and the next time you open the preferences, it has gone:
![image](https://github.com/Mudlet/Mudlet/assets/6163092/2b248c3f-9c82-4459-955f-2e4dd988bbdf)

To make the current `development` branch code work with a "migrated" Discord account it is necessary to both clear the discriminator AND change the user name to its new format (lower case ASCII letters, digits, `_` or `.` ONLY) as per https://support.discord.com/hc/en-us/articles/12620128861463-New-Usernames-Display-Names .
![image](https://github.com/Mudlet/Mudlet/assets/6163092/fc68e35a-d8ea-435e-9972-c923ae18d0d9)

*then* it will work:
![image](https://github.com/Mudlet/Mudlet/assets/6163092/6bd7acef-d1fb-4f05-b6ca-3f37b95ca0a4)

Importantly - entering `0000` as the discriminator:
![image](https://github.com/Mudlet/Mudlet/assets/6163092/850d84b2-10ca-4030-b968-94987b755c3f)
does **not** work:
![image](https://github.com/Mudlet/Mudlet/assets/6163092/f30790a5-4ea6-4a67-83f1-13336accbb0b)

That is the behaviour of the current code base, but with this PR warnings/explanations are given each time the profile is started should there be a problem with the user name or the discriminator going forward - and, using my old details these look like:
![image](https://github.com/Mudlet/Mudlet/assets/6163092/cef58aa6-7a90-4aa2-81a0-5eb7ddecd90d)

I notice in producing this screen shot that there is an unwanted line-feed in the "name" warning and an unwanted  single quote in the "discriminator" one - I'll fix that as soon as possible - however once the changes are made by the end-user the warning goes away and the single digit zero discriminator is allowed and persists in the settings:
![image](https://github.com/Mudlet/Mudlet/assets/6163092/a5430580-fb5f-47e7-8140-d32188ec1f48)

Note: this username change will happen for all Discord users but I do not know yet how far it has progressed and the user cannot make the change until Discord notifies them and - should this PR go in they will keep getting the warning(s) until they can - but they do explain what they will need to do ***in Mudlet*** to adapt the settings in it  to them.

SlySven committed:
Revise: fix a couple of typos in warning messages

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Fix_allow_singleZeroDigitDiscordDiscriminator

SlySven committed:
Revise: fix another typo I put into the previous commit

:roll_eyes:

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
@Kebap

Kebap commented:
So to test this, a user needs
- A game which supports Discord rich presence
- (At least) two separate Discord accounts
- They must use this Mudlet feature with their old Discord user number
- The new Discord user numbering system must be activated for their user
- Did I miss anything?

Do we have any reports of this happening to anyone in real life, quite yet?
I mean somebody other than the author of this PR

I don't use this Mudlet feature myself, so I am not affected.
To test this, I think I must first create a second Discord user.
Then again, SlySven did already test and document himself.

Another thing that bothers me. There is this warning shown to users:

> That is the behaviour of the current code base, but with this PR warnings/explanations are given each time the profile is started should there be a problem with the user name or the discriminator going forward - and, using my old details these look like: 
> ![grafik](https://github.com/Mudlet/Mudlet/assets/117238/75adb115-d45c-496c-a0da-37b5be0fb133)

- This text is 2x8 lines long wall of text, should be 2-3 max
- This text is shown at every start of every Mudlet profile affected
- These texts for name and discriminator even have a lot of repetition
- This text is also already shown while the users can't do anything against it yet
- The content, describing on which preferences/settings tab a setting is located, is unusual at best

SlySven commented:
Actually my secondary identity/account on Discord got a message last week - I have until 2024/03/04 to rename that one else Discord will do it for me - and thus remove the "discriminator" on that one. TBH perhaps we should just discard the setting and strip it from any profiles that have anything stored for it... and also advise users of existing versions that they will need to clear that setting in their current versions of Mudlet...

![Screenshot_20240228_193642](https://github.com/Mudlet/Mudlet/assets/6163092/131b8640-6f27-4ad7-862a-da680457defe)


:man_shrugging:

SlySven committed:
Merge branch 'development' into Fix_allow_singleZeroDigitDiscordDiscriminator

vadi2 commented:
> TBH perhaps we should just discard the setting and strip it from any profiles that have anything stored for it... 

That would be good 👍 

> and also advise users of existing versions that they will need to clear that setting in their current versions of Mudlet...

We could do that in our #announcements without an @everyone tag, since we know few people filled it in.

SlySven committed:
Fix: remove Discord discriminator completely

Prints a warning if one is found in the game save file - advising that it
is no longer saved (so will be forgotten when the session is saved).

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
We haven't had any players ask about this for a year... I think the community is not really interested in this. Lets revisit when something should change.
